### PR TITLE
Add :order_by option to list and list_by

### PR DESCRIPTION
**Pull request by palexanderm**
Currently, the default sort order in `list` and `list_by` is according to `:inserted_at`. With adding an `:order_by` option, `list` and `list_by` could now be sorted according to any key in the model. The syntax would be:
- `Model.list(order_by: :name)`
- `Model.list(order_by: [desc: :name]`
- `Model.list_by(type: 1, order_by: [desc: :field1, asc: :field2]`

### CODE REVIEW

File: lib/one_model.ex [hunk 1]
```diff
@@ -141,38 +141,40 @@ defmodule OneModel do
...
+      * `order_by: atom | list`
       """
       @spec list(Keword.t()) :: [Struct.t()]
       def list(opts \\ []) do
         if preload = opts[:preload] do
           @schema
           |> preload(^preload)
-          |> order_by(asc: :inserted_at)
-          |> @repo.all
+          |> do_order(opts[:order_by])
```
Comment (smpallen99) [inline lib/one_model.ex:154]: I don't like the the code duplication here. U should be able to call do_preload. 

```
@schema
|> do_preload(opts[:preload])
|> do_order(...)
|> @repo.all()
```

**Review by smpallen99 — APPROVED** [review]
Small refactor in list

**Comment by smpallen99** [comment]
Paul, I don't understand why there are conflicts. Its almost as if u did not base ur branch off of master.

**Comment by Sinc63** [comment]
> Paul, I don't understand why there are conflicts. Its almost as if u did not base ur branch off of master.

He didn't. The master branch has the dialyzer fixes which we integrated for release 4. Those require some corresponding code change in the InfinityOne code, which weren't made in release 3. Release 3 won't compile with the master content, so we deliberately branched this from the previous content and will be using this branch for R3 and will have to merge this to master to use it for R4.

File: lib/one_model.ex [hunk 2]
```diff
@@ -141,38 +141,40 @@ defmodule OneModel do
...
 
       Pass a list of preloads with the `:preload` key.
 
       ## Examples
 
-          #{@schema}.list_by field1: value1, field2: field2, preload: [:association]
+          #{@schema}.list_by(field1: value1, field2: value2, preload: [:association])
       """
-      @spec list_by(Keyword.t()) :: List.t()
+      @spec list_by(Keword.t()) :: [Struct.t()]
```
Comment (Sinc63) [inline lib/one_model.ex:174]: I'm not sure why your changes are showing this change. It's almost as if you based this from master and then put back the old changes. This line should already be like this in the pre-dialyzer version I think. Weird. But let's not revert this to the bad version, regardless of the way it got here. Restore the line with the spelling fix and the List.t.

File: lib/one_model.ex [hunk 3]
```diff
@@ -181,20 +183,31 @@ defmodule OneModel do
       """
       @spec list_by_query(Keyword.t(), Keyword.t()) :: [Struct.t()]
       def list_by_query(params, opts \\ []) do
-        {preload, params} = Keyword.pop(params, :preload, [])
-        preload = if opts[:preload] == false or preload == [], do: false, else: preload
+        {pre_ord, params} = Keyword.split(params, ~w(preload order_by)a)
+
+        preload =
+          if opts[:preload] == false or pre_ord[:preload] == [],
```
Comment (Sinc63) [inline lib/one_model.ex:189]: The second check should be `is_nil(pre_ord:preload])`. If there is no preload value the list will be empty, but indexing by an undefined keyword gets nil.

You need to add some tests for the new option and make sure there are tests of preloads included.

File: lib/one_model.ex (hunk 1 above)
Comment (palexanderm) [reply lib/one_model.ex:154]: Good call. Not sure how I missed that.

File: lib/one_model.ex (hunk 2 above)
Comment (palexanderm) [reply lib/one_model.ex:174]: In discussion with Steve, we came to the conclusion that having a separate branch for iOne3 would be unnecessary. I've created a new branch based off master, not excluding the most recent commit, and have made the appropriate changes within the iOne code to resolve the compile errors with the types. This way iOne 3 and 4 will both use the master branch. I will be opening a new PR for that branch.

File: lib/one_model.ex (hunk 3 above)
Comment (palexanderm) [reply lib/one_model.ex:189]: Corrected.

**Comment by palexanderm** [comment]
> Paul, I don't understand why there are conflicts. Its almost as if u did not base ur branch off of master.

**Pull request closed without merging**